src/wallet/wallet.cpp

```diff
@@ -2709,41 +2712,54 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
-                    // Reduce fee to only the needed amount if we have change
-                    // output to increase.  This prevents potential overpayment
-                    // in fees if the coins selected to meet nFeeNeeded result
-                    // in a transaction that requires less fee than the prior
-                    // iteration.
-                    // TODO: The case where nSubtractFeeFromAmount > 0 remains
-                    // to be addressed because it requires returning the fee to
-                    // the payees and not the change output.
-                    // TODO: The case where there is no change output remains
-                    // to be addressed so we avoid creating too small an output.
-                    if (nFeeRet > nFeeNeeded && nChangePosInOut != -1 && nSubtractFeeFromAmount == 0) {
-                        CAmount extraFeePaid = nFeeRet - nFeeNeeded;
-                        std::vector<CTxOut>::iterator change_position = txNew.vout.begin()+nChangePosInOut;
-                        change_position->nValue += extraFeePaid;
-                        nFeeRet -= extraFeePaid;
+                // Modify fee to the needed amount from the change
+                // output, or recipients in the case of subtractFeeFromAmount.
+                // This prevents potential overpayment in fees if
+                // the coins selected to meet nFeeNeeded result
+                // in a transaction that requires less fee than the prior
+                // iteration.
+                CAmount excessFee = nFeeRet - nFeeNeeded;
+
+                if (nSubtractFeeFromAmount == 0) {
+                    // Take fee excess
+                    txNew.vout[nChangePosInOut].nValue += excessFee;;
+                    nFeeRet -= excessFee;
+
+                    // Negative change value means include more fee and try again.
+                    if (txNew.vout[nChangePosInOut].nValue < 0) {
+                        reservekey.ReturnKey();
+                        nFeeRet = nFeeNeeded;
+                        continue;
                     }
-                    break; // Done, enough fee included.
+                    // Drop change if dust
+                    // TODO replace with economical change calc
+                    else if (IsDust(txNew.vout[nChangePosInOut], ::dustRelayFee)) {
+                        nFeeRet += txNew.vout[nChangePosInOut].nValue;
+                        txNew.vout.erase(txNew.vout.begin() + nChangePosInOut);
+                        nChangePosInOut = -1;
+                        reservekey.ReturnKey();
+                    // If larger than dust, but still small, increase fee target and try again
+                    } else if (txNew.vout[nChangePosInOut].nValue < MIN_FINAL_CHANGE) {
+                        // Save this transaction, use if we cannot get large-enough change
+                        if (!have_cached_txn) {
+                            tx_cached = txNew;
```

Comment:
`have_cached_txn` should probably be set to true here as it is never actually set anywhere